framework/src/main/java/org/checkerframework/common/accumulation/AccumulationTransfer.java

```diff
@@ -0,0 +1,125 @@
+        super(analysis);
+        typeFactory = (AccumulationAnnotatedTypeFactory) analysis.getTypeFactory();
+    }
+
+    /**
+     * Updates the estimate of how many things {@code node} has accumulated.
+     *
+     * <p>If the node is an invocation of a method that returns its receiver, then its receiver's
+     * type will also be updated. In a chain of method calls, this process will continue backward as
+     * long as each receiver is itself a receiver-returning method invocation.
+     *
+     * <p>For example, suppose {@code node} is the expression {@code a.b().c()}, the new value
+     * (added by the accumulation analysis because of the {@code .c()} call) is "foo", and b and c
+     * return their receiver. This method will directly update the estimate of {@code a.b().c()} to
+     * include "foo". In addition, the estimates for the expressions {@code a.b()} and {@code a}
+     * would have their estimates updated to include "foo", because c and b (respectively) return
+     * their receivers. Note that due to what kind of values can be held in the store, this
+     * information is lost outside the method chain. That is, the returns-receiver propagated
+     * information is lost outside the expression in which the returns-receiver method invocations
+     * are nested.
+     *
+     * <p>As a concrete example, consider the Called Methods accumulation checker: if {@code build}
+     * requires a, b, and c to be called, then {@code foo.a().b().c().build();} will typecheck (they
+     * are in one fluent method chain), but {@code foo.a().b().c(); foo.build();} will not -- the
+     * store does not keep the information that a, b, and c have been called outside the chain.
+     * {@code foo}'s type will be {@code CalledMethods("a")}, because only {@code a()} was called
+     * directly on {@code foo}. For such code to typecheck, the Called Methods accumulation checker
+     * uses an additional rule: the return type of a receiver-returning method {@code rr()} is
+     * {@code CalledMethods("rr")}. This rule is implemented directly in the {@link
+     * org.checkerframework.framework.type.treeannotator.TreeAnnotator} subclass defined in the
+     * Called Methods type factory.
+     *
+     * @param node the node whose estimate should be expanded
+     * @param result the transfer result containing the store to be modified
+     * @param values the new accumulation values
+     */
+    public void accumulate(Node node, TransferResult<CFValue, CFStore> result, String... values) {
+        List<String> valuesAsList = Arrays.asList(values);
+        // If dataflow has already recorded information about the target, fetch it and integrate
+        // it into the list of values in the new annotation.
+        CFValue flowValue = result.getResultValue();
+        if (flowValue != null) {
+            Set<AnnotationMirror> flowAnnos = flowValue.getAnnotations();
+            for (AnnotationMirror anno : flowAnnos) {
```

Comment:
Is this `for` loop guaranteed to execute exactly 0 or 1 times?  Should it assert that about the size of `flowAnnos`?  I think that each one must be either the accumulator or bottom; is that correct?  Is that worth verifying?

Reply:
> Is this for loop guaranteed to execute exactly 0 or 1 times

Yes. I added an assertion.

> I think that each one must be either the accumulator or bottom; is that correct?

Yes. My original version of this code included a test for it, but it was redundant with checking that the result of the call to `getValueOfAnnotationWithStringArgument` is non-null, so I removed it.